### PR TITLE
[SPARK-16803] [SQL] SaveAsTable does not work when source DataFrame is built on a Hive Table

### PR DESCRIPTION
**Pull request by gatorsmile**
### What changes were proposed in this pull request?

In Spark 2.0, `SaveAsTable` does not work when source DataFrame is built on a Hive Table, but Spark 1.6 works.

**Spark 1.6**

``` Scala
scala> sql("create table sample.sample stored as SEQUENCEFILE as select 1 as key, 'abc' as value")
res2: org.apache.spark.sql.DataFrame = []

scala> val df = sql("select key, value as value from sample.sample")
df: org.apache.spark.sql.DataFrame = [key: int, value: string]

scala> df.write.mode("append").saveAsTable("sample.sample")

scala> sql("select * from sample.sample").show()
+---+-----+
|key|value|
+---+-----+
|  1|  abc|
|  1|  abc|
+---+-----+
```

**Spark 2.0**

``` Scala
scala> df.write.mode("append").saveAsTable("sample.sample")
org.apache.spark.sql.AnalysisException: Saving data in MetastoreRelation sample, sample
 is not supported.;
```

This PR is to provide a support with by-name resolution. In 1.6, it is by-position resolution. The previous behavior is wrong. We need to adjust the order.
### How was this patch tested?

Test cases are added

### CODE REVIEW

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala [hunk 1]
```diff
@@ -1473,6 +1473,94 @@ class SQLQuerySuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
...
+
+      df.write.insertInto(tableName)
+      checkAnswer(
+        sql(s"SELECT * FROM $tableName"),
+        Row(1, "abc") :: Row(1, "abc") :: Nil
+      )
+    }
+  }
+
+  test("saveAsTable - format Hive") {
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: @cloud-fan Should we disable the format `hive` in DataFrameWriter? If this is supported, we might need a separate PR to cover all the modes. Thanks!

**Comment by gatorsmile** [comment]
cc @cloud-fan @yhuai

**Comment by gatorsmile** [comment]
retest this please

**Comment by cloud-fan** [comment]
hi @gatorsmile do you have time to work on it? I'd like to get this into 2.1

**Comment by gatorsmile** [comment]
Sure, will finish this tomorrow.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala [hunk 2]
```diff
@@ -89,6 +89,22 @@ case class AnalyzeCreateTable(sparkSession: SparkSession) extends Rule[LogicalPl
       }
       c
 
+    case c @ CreateTable(tableDesc, mode, Some(query))
+        if mode == SaveMode.Append && isHiveSerdeTable(tableDesc.identifier) =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: can we consolidate hive and data source table here?

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: Let me try another way to fix it. Will submit a new PR.

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: uh... Actually, I found a bug in our write path.

**Pull request closed without merging**